File: AZ-204_Implement_Secure_Azure_Solutions/code/resource_mgmnt.cs
using System;
using System.Threading.Tasks;
using Microsoft.Azure.Management.Network;
using Microsoft.Azure.Management.Network.Models;
using Microsoft.Azure.Management.KeyVault;
using Microsoft.Azure.Management.KeyVault.Models;
using Microsoft.Azure.Management.ResourceManager;
using Microsoft.Azure.Management.ResourceManager.Models;
using Microsoft.Rest;

class Program
{
    static async Task Main(string[] args)
    {
        string subscriptionId = "<Your Subscription ID>";
        string resourceGroupName = "<Your Resource Group Name>";
        string location = "<Your Resource Group Location>";

        // Authenticate using your Azure credentials
        var credentials = SdkContext.AzureCredentialsFactory.FromFile("<Your Azure Credentials File>");

        // Create a Network Security Group
        var networkSecurityGroup = await CreateNetworkSecurityGroup(credentials, subscriptionId, resourceGroupName, location);
        Console.WriteLine("Network Security Group created successfully!");

        // Update the Network Security Group
        await UpdateNetworkSecurityGroup(credentials, subscriptionId, resourceGroupName, networkSecurityGroup.Id);
        Console.WriteLine("Network Security Group updated successfully!");

        // Delete the Network Security Group
        await DeleteNetworkSecurityGroup(credentials, subscriptionId, resourceGroupName, networkSecurityGroup.Id);
        Console.WriteLine("Network Security Group deleted successfully!");

        // Create a Key Vault
        var keyVault = await CreateKeyVault(credentials, subscriptionId, resourceGroupName, location);
        Console.WriteLine("Key Vault created successfully!");

        // Update the Key Vault
        await UpdateKeyVault(credentials, subscriptionId, resourceGroupName, keyVault.Id);
        Console.WriteLine("Key Vault updated successfully!");

        // Delete the Key Vault
        await DeleteKeyVault(credentials, subscriptionId, resourceGroupName, keyVault.Id);
        Console.WriteLine("Key Vault deleted successfully!");
    }

    static async Task<NetworkSecurityGroup> CreateNetworkSecurityGroup(ServiceClientCredentials credentials, string subscriptionId, string resourceGroupName, string location)
    {
        var networkClient = new NetworkManagementClient(credentials) { SubscriptionId = subscriptionId };

        var nsgParams = new NetworkSecurityGroup()
        {
            Location = location,
            SecurityRules = new List<SecurityRule>()
            {
                new SecurityRule()
                {
                    Name = "Allow-SSH",
                    Protocol = SecurityRuleProtocol.Tcp,
                    SourcePortRange = "*",
                    DestinationPortRange = "22",
                    SourceAddressPrefix = "*",
                    DestinationAddressPrefix = "*",
                    Access = SecurityRuleAccess.Allow,
                    Priority = 100,
                    Direction = SecurityRuleDirection.Inbound
                }
            }
        };

        return await networkClient.NetworkSecurityGroups.CreateOrUpdateAsync(resourceGroupName, "MyNSG", nsgParams);
    }

    static async Task UpdateNetworkSecurityGroup(ServiceClientCredentials credentials, string subscriptionId, string resourceGroupName, string nsgId)
    {
        var networkClient = new NetworkManagementClient(credentials) { SubscriptionId = subscriptionId };

        var nsg = await networkClient.NetworkSecurityGroups.GetAsync(resourceGroupName, "MyNSG");
        nsg.SecurityRules.Add(new SecurityRule()
        {
            Name = "Allow-HTTP",
            Protocol = SecurityRuleProtocol.Tcp,
            SourcePortRange = "*",
            DestinationPortRange = "80",
            SourceAddressPrefix = "*",
            DestinationAddressPrefix = "*",
            Access = SecurityRuleAccess.Allow,
            Priority = 200,
            Direction = SecurityRuleDirection.Inbound
        });

        await networkClient.NetworkSecurityGroups.CreateOrUpdateAsync(resourceGroupName, "MyNSG", nsg);
    }

    static async Task DeleteNetworkSecurityGroup(ServiceClientCredentials credentials, string subscriptionId, string resourceGroupName, string nsgId)
    {
        var networkClient = new NetworkManagementClient(credentials) { SubscriptionId = subscriptionId };

        await networkClient.NetworkSecurityGroups.DeleteAsync(resourceGroupName, "MyNSG");
    }

    static async Task<Vault> CreateKeyVault(ServiceClientCredentials credentials, string subscriptionId, string resourceGroupName, string location)
    {
        var keyVaultClient = new KeyVaultManagementClient(credentials) { SubscriptionId = subscriptionId };

        var vaultParams = new VaultCreateOrUpdateParameters()
        {
            Location = location,
            Properties = new VaultProperties()
            {
                Sku = new Sku()
                {
                    Name = SkuName.Standard
                },
                TenantId = "<Your Tenant ID>",
                AccessPolicies = new List<AccessPolicyEntry>()
                {
                    new AccessPolicyEntry()
                    {
                        TenantId = "<Your Tenant ID>",
                        ObjectId = "<Your Object ID>",
                        Permissions = new Permissions()
                        {
                            Keys = new List<string>() { "all" },
                            Secrets = new List<string>() { "all" }
                        }
                    }
                }
            }
        };

        return await keyVaultClient.Vaults.CreateOrUpdateAsync(resourceGroupName, "MyKeyVault", vaultParams);
    }

    static async Task UpdateKeyVault(ServiceClientCredentials credentials, string subscriptionId, string resourceGroupName, string vaultId)
    {
        var keyVaultClient = new KeyVaultManagementClient(credentials) { SubscriptionId = subscriptionId };

        var vault = await keyVaultClient.Vaults.GetAsync(resourceGroupName, "MyKeyVault");
        vault.Properties.Sku.Name = SkuName.Premium;

        await keyVaultClient.Vaults.CreateOrUpdateAsync(resourceGroupName, "MyKeyVault", vault);
    }

    static async Task DeleteKeyVault(ServiceClientCredentials credentials, string subscriptionId, string resourceGroupName, string vaultId)
    {
        var keyVaultClient = new KeyVaultManagementClient(credentials) { SubscriptionId = subscriptionId };

        await keyVaultClient.Vaults.DeleteAsync(resourceGroupName, "MyKeyVault");
    }
}

File: AZ-204_Implement_Secure_Azure_Solutions/code/resource_mgmnt.py
from azure.identity import DefaultAzureCredential
from azure.mgmt.network import NetworkManagementClient
from azure.mgmt.keyvault import KeyVaultManagementClient
from azure.mgmt.resource import ResourceManagementClient

# Set your Azure subscription ID and resource group name
subscription_id = 'your_subscription_id'
resource_group_name = 'your_resource_group_name'

# Set your Azure credentials
credential = DefaultAzureCredential()

# Create a Network Security Group
network_client = NetworkManagementClient(credential, subscription_id)
nsg_params = {
    'location': 'westus',
    'security_rules': []
}
nsg = network_client.network_security_groups.create_or_update(resource_group_name, 'my_nsg', nsg_params)

# Create a Key Vault
keyvault_client = KeyVaultManagementClient(credential, subscription_id)
keyvault_params = {
    'location': 'westus',
    'properties': {
        'sku': {
            'name': 'standard'
        },
        'tenant_id': 'your_tenant_id',
        'access_policies': [],
        'enabled_for_disk_encryption': False,
        'enabled_for_deployment': True,
        'enabled_for_template_deployment': True
    }
}
keyvault = keyvault_client.vaults.create_or_update(resource_group_name, 'my_keyvault', keyvault_params)

# Update the NSG
nsg_params['security_rules'].append({
    'name': 'allow_ssh',
    'protocol': 'Tcp',
    'source_port_range': '*',
    'destination_port_range': '22',
    'source_address_prefix': '*',
    'destination_address_prefix': '*',
    'access': 'Allow',
    'priority': 100,
    'direction': 'Inbound'
})
nsg = network_client.network_security_groups.create_or_update(resource_group_name, 'my_nsg', nsg_params)

# Update the Key Vault
keyvault_params['properties']['access_policies'].append({
    'tenant_id': 'your_tenant_id',
    'object_id': 'your_object_id',
    'permissions': {
        'keys': ['get', 'list', 'create', 'update', 'import', 'delete', 'backup', 'restore', 'recover', 'purge'],
        'secrets': ['get', 'list', 'set', 'delete', 'backup', 'restore', 'recover', 'purge'],
        'certificates': ['get', 'list', 'delete', 'create', 'import', 'update', 'managecontacts', 'getissuers', 'listissuers', 'setissuers', 'deleteissuers', 'manageissuers', 'recover', 'purge']
    }
})
keyvault = keyvault_client.vaults.create_or_update(resource_group_name, 'my_keyvault', keyvault_params)

# Delete the NSG
network_client.network_security_groups.begin_delete(resource_group_name, 'my_nsg').wait()

# Delete the Key Vault
keyvault_client.vaults.begin_delete(resource_group_name, 'my_keyvault').wait()